<!--This file launches either a bag or a simulation and the point-cloud converter.--> 
<!--  Newtorks can be launched through the desiccation_cnn package -->
<!--  -->
<launch>
  <!-- <arg name="debug" default="true"/> -->
  <!-- Simulation arguments -->
  <arg name="world" default="serf"/>
  <!-- Values for the camera calibration file -->
  <arg name="pcl" default="false"/>
  <arg name="delay" default="10"/>
  <arg name="respawn" default="true"/>
  <arg name="cnn" default="false"/>
  <arg name="rosbag" default="false"/>
  <arg unless="$(arg cnn)" name="out_width" default="256"/> <!-- 1920 / 256 / 128 --> 
  <arg unless="$(arg cnn)" name="out_height" default="192"/> <!-- 1080 / 192 / 96 -->
  <arg if="$(arg cnn)" name="out_width" default="128"/> <!-- 1920 / 256 / 128 --> 
  <arg if="$(arg cnn)" name="out_height" default="96"/> <!-- 1080 / 192 / 96 -->
  <arg name="static_tf" default="False"/>
  <!-- Visualisation -->
  <arg name="rviz" default="True"/>
  <!-- Frames -->
  <arg name="pcl_topic" default="pcl_plane"/>
  <arg name="world_frame" default="map"/>
  <!-- Arguments for the quadmap -->
  <arg name="quadmap" default="false"/>
  <arg name="max_depth" default="10"/>
  <arg name="qt_scale" default="70"/>
  <arg name="low" default="-40"/>  
  <!-- Argument for tf throttle -->
  <arg name="throttle" default="false"/>
  <arg unless="$(arg throttle)" name="throttle_topic" value=""/>
  <arg if="$(arg throttle)" name="throttle_topic" value="/tf_throttle:=/tf"/>
  

  <!-- Setting the names depending on image / processed  -->
  <arg name="camera_name" default="/hbv_1615"/>
  <arg unless="$(arg cnn)" name="image_topic" default="$(arg camera_name)/image_color"/>
  <arg if="$(arg cnn)" name="image_topic" default="$(arg camera_name)/detection"/>
  <arg name="camera_frame" default="hbv_1615_link"/>
  <arg name="info_topic" default="$(arg camera_name)/camera_info"/>
  
  <!-- Simulation -->
  <include file="$(find desiccation_cnn)/launch/simulation.launch">
      <arg name="world" value="$(arg world)"/>
  </include>

  <!-- camera HITL - a lot of hardcoded things in there -->
  <node pkg="desiccation_cnn" type="oakD_hitl.py" name="oakD_hitl" output="screen" respawn="false" respawn_delay="0"/>

  <!-- Set the HITL parameter if we have pcl:=true - to call the delay in the cpp file -->
  <param if="$(arg pcl)" name="delay" type="int" value="$(arg delay)"/>
  <!-- The pcl file - uses cascaded calibration file -->
  <include if="$(arg pcl)" file="$(find learning_image_geometry)/launch/pcl.launch">
    <arg name="camera_name" value="$(arg camera_name)"/>
    <arg name="out_width" value="$(arg out_width)"/>
    <arg name="out_height" value="$(arg out_height)"/>
    <arg name="static_tf" value="$(arg static_tf)"/>
    <arg name="camera_frame" value="$(arg camera_frame)"/>
    <arg name="pcl_topic" value="$(arg pcl_topic)"/>
    <arg name="world_frame" value="$(arg world_frame)"/>
    <arg name="info_topic" value="$(arg info_topic)"/>
    <arg name="image_topic" value="$(arg image_topic)"/>
    <arg name="respawn" value="$(arg respawn)"/>
  </include>
  
  <!-- RVIZ -->
  <node if="$(arg rviz)" pkg="rviz" type="rviz" output="screen" name="rviz" respawn="true" args="-d $(find quadmap)/config/rviz_hitl.rviz"/>

  <!-- New transform to make all the values positive -->
  <!-- <node pkg="tf" type="static_transform_publisher" name="tf_new" args="0 0 0 1.57 0 0 positive_local_origin local_origin 30"/> -->

  <!-- The python node running the quadtree -->
  <param name="qt_scale" type="int" value="$(arg qt_scale)"/>
  <param name="max_depth" type="int" value="$(arg max_depth)"/>
  <param name="low" type="int" value="$(arg low)"/>
  <node if="$(arg quadmap)" pkg="quadmap" type="quadtree_node.py" name="quadtree_node" output="screen"/>

  <!-- Rosbag Record -->
  <!-- TODO: Continue here -->
  <!-- ! since we only ever need the pcl, we can just run the pcl recorder here -->
  <!-- ? what is the file size of the rosbag if we only record the pcl - ! roughly 1 GB with the PCL for the quick missions! -->
  <node if="$(arg rosbag)" pkg="rosbag" type="record" name="recorder" args="record -o /home/nicolas/rosbag/sim/hitl /$(arg pcl_topic) /tf"/>
</launch>
